Projects/UOContent/Mobiles/AI/BaseAI.cs

```diff
@@ -144,172 +146,212 @@ public ActionType Action
         get => m_Action;
         set
         {
-            m_Action = value;
-            OnActionChanged();
+            if (m_Action != value)
+            {
+                m_Action = value;
+                OnActionChanged();
+            }
         }
     }
 
     public long NextMove { get; set; }
 
+    // OSI, DetectHidden >= 50.0?
     public virtual bool CanDetectHidden => m_Mobile.Skills.DetectHidden.Value > 0;
-
+    
     public virtual bool WasNamed(string speech)
```

Comment:
We should assume `WasNamed` is idempotent against the same speech. This means all of the uses for each switch/loop/sub-function will return the same value. We can refactor by reducing the number of times we call WasNamed to once and use that bool value where we need it.

For loops where all commands require WasNamed to be true, we can do a single check before the loop too.